EC-88 - Sender should record sync message sent to activeMQ

lccpinto opened the pull request:
Recreating the pull request after repo cleaning process.
See: EC-153 - Fulfill GitHub requested changes to avoid project disabling

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:
```diff
@@ -0,0 +1,104 @@
+	private String tableName;
+	
+	@NotNull
+	@Column(length = 255, nullable = false, updatable = false)
```
This should not be updatable too

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:
```diff
@@ -0,0 +1,104 @@
+	private String identifier;
+	
+	@NotNull
+	@Column(length = 1, nullable = false, updatable = false)
```
This should not be updatable too

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:
```diff
@@ -0,0 +1,104 @@
+	private String requestUuid;
+	
+	@NotNull
+	@Column(length = 1, nullable = false)
```
You don't need to specify length for a boolean column

lccpinto replied:
Right, fixed.

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:
```diff
@@ -0,0 +1,104 @@
+	private boolean sent;
+	
+	@NotNull
+	@Column(length = 1, nullable = false, updatable = false)
```
You don't need to specify length for a boolean column

lccpinto replied:
Fixed.

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SyncResponse.java:
```diff
@@ -0,0 +1,80 @@
+	@NotNull
+	@Column(name = "message_uuid", length = 38, nullable = false, unique = true, updatable = false)
+	@Access(AccessType.FIELD)
+	private String messageUuid;
```
I think we should not make message_uuid unique because I can imagine a scenario where the receiver might send a response for the same sender message more than once e.g. on a retry attempt in case something went wrong when actually the first attempt went through.

lccpinto replied:
Alright. I've made the change.

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SyncResponse.java:
```diff
@@ -0,0 +1,80 @@
+	private String messageUuid;
+	
+	@NotNull
+	@Column(name = "date_sent", precision = 3, nullable = false, updatable = false)
```
I don't think you need to set the precision

lccpinto replied:
Fixed.

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SyncResponse.java:
```diff
@@ -0,0 +1,80 @@
+	
+	@NotNull
+	@Enumerated(EnumType.STRING)
+	@Column(nullable = false, length = 10)
```
I think 10 might be too small just in case we add future values, may be try 20

lccpinto replied:
Right. Changed.

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SyncResponse.java:
```diff
@@ -0,0 +1,80 @@
+	private SyncResponseStatus status;
+	
+	public SyncResponse(@NotNull String messageUuid, @NotNull Date dateSent) {
+		super();
```
does the super class no anything actually? This might be redundant

lccpinto replied:
No. This was put by the Eclipse constructor generator.
Removed.

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SyncResponse.java:
```diff
@@ -0,0 +1,80 @@
+	@Access(AccessType.FIELD)
+	private SyncResponseStatus status;
+	
+	public SyncResponse(@NotNull String messageUuid, @NotNull Date dateSent) {
```
This being a persistent class, you need to arg constructor

wluyima replied:
Do you use this other constructor anywhere? If not, you might as well get rid of it

wluyima replied:
@lccpinto did you see this comment?

lccpinto replied:
Sorry @wluyima, I didn't see because it was marked as "resolved" and I was hiding the resolved chats.
I was using this constructor, but I did a refactor to not use a constructor at all, to be aligned with the rest of the code.

wluyima commented on app-api/src/main/resources/liquibase-sender.xml:
```diff
@@ -250,5 +250,74 @@
+                <constraints nullable="false" unique="true"/>
+            </column>
+            <column name="request_uuid" type="VARCHAR(38)">
+                <constraints unique="true"/>
```
Remove unique constraint

lccpinto replied:
This is the sender_sync_message table. There is no chance to have duplicates here, since the request_uuid is generated per request.
I think that you wanna tell to remove in the sync_response table, where we could have multiple responses for the same request_uuid. Right?

wluyima replied:
Recall that the receiver can request an entity from the a remote site, usually that request uuid remains is what gets copied over to this column, recall that? So it's not true that the sender always generates a new one

lccpinto replied:
Sorry, I'd misunderstood. I was switching concepts between **request_uuid** and **message_uuid**.
Fixed.

wluyima commented on app-api/src/main/resources/liquibase-sender.xml:
```diff
@@ -250,5 +250,74 @@
+            <column name="sent" type="BOOLEAN">
+                <constraints nullable="false" />
+            </column>
+            <column name="date_created" type="TIMESTAMP(3)">
```
I think this can be a regular date

lccpinto replied:
Changed.

wluyima commented on app-api/src/main/resources/liquibase-sender.xml:
```diff
@@ -250,5 +250,74 @@
+    <changeSet author="lpinto" id="20220323-1545">
+        <preConditions onFail="MARK_RAN">
+            <not>
+                <tableExists tableName="sync_response" />
```
Rename to sender_sync_response

lccpinto replied:
Renamed. Should we rename the java entities and routes also?

SyncResponse.java
sync-response-processor.xml

wluyima commented on distribution/sender/routes/active-mq-publisher.xml:
```diff
@@ -0,0 +1,132 @@
+    <route id="active-mq-publisher" errorHandlerRef="outBoundErrorHandler">
+    
+        <!-- TODO(luis.pinto) Make the delay configurable -->
+        <from uri="scheduler:active_mq_publisher?initialDelay=15000&amp;delay=60000" />
```
Initial delay should be may be 2minutes I think, it gives the application and other more important routes higher priority in the bootstrap process.

lccpinto replied:
Fixed. Delays moved to configurations.

wluyima commented on distribution/sender/routes/active-mq-publisher.xml:
```diff
@@ -0,0 +1,132 @@
+
+        <log message="Fetching next 1000 sender sync messages" loggingLevel="DEBUG" />
+
+        <toD uri="jpa:SenderSyncMessage?query=SELECT m FROM SenderSyncMessage m WHERE m.sent = false ORDER BY m.dateCreated ASC, m.id ASC&amp;maximumResults=1000" />
```
With the number of processor routes increasing, may be we should reduce the limit to 500 to cut on the memory footprint

lccpinto replied:
Changed. Configuration created.

wluyima commented on distribution/sender/routes/active-mq-publisher.xml:
```diff
@@ -0,0 +1,132 @@
+                    </choice>
+            
+                    <when>
+                        <simple>${body} == null</simple>
```
You have altered some logic here, the condition should be,

`${body} != null || ${exchangeProperty.event.operation} == 'r'`

wluyima replied:
This is because the entity might no longer exist in the database

lccpinto replied:
You're totally right, I was misunderstanding the flow. I've fixed it. Thank you.

wluyima commented on distribution/sender/routes/active-mq-publisher.xml:
```diff
@@ -0,0 +1,132 @@
+    
+                    <toD uri="{{camel.output.endpoint}}" />
+    
+                    <log message="Entity sent! Updating request status to sent." />
```
Some changes have happened in master in the original route, this is now logged at debug

lccpinto replied:
Alright. Changed.

wluyima commented on distribution/sender/routes/db-sync-route.xml:
```diff
@@ -4,100 +4,61 @@
 
         <log message="Start db-sync-route: ${body}" loggingLevel="DEBUG" />
-
+        
```
Why do you need this property? Try as much to avoid exchange properties, they can easily be duplicated in routes which can cause obscure bugs.

lccpinto replied:
I have removed the "isValid" exchange property during a refactor. Thanks.

wluyima commented on distribution/sender/routes/request-consumer.xml:
```diff
@@ -1,45 +0,0 @@
-<routes xmlns="http://camel.apache.org/schema/spring">
```
Why did you delete this route?

lccpinto replied:
Do you remember the refactor that you've asked to be done in the first pull request?
There, we noticed that the **request-consumer** and **sender-message-consumer** would going to consume the same queue (openmrs.sync.{{db-sync.senderId}}).
So, you sugested to merge this two routes in one. Now, the logic that were in the **request-consumer** route was moved to **sender-message-consumer** route.
Thanks.

lccpinto replied:
See the message below from the first pull request:

_wluyima
 Collaborator
"I think we should use a single route to process all incoming activeMQ messages i.e. request-
consumer should be renamed to something else generic like sender-message-consumer
because it is a single queue we are reading from and then based on the message contents we
save the incoming message to the respective table i.e. sender_sync_request or
sync_response and then this route should be actually processing messages from the
sync_response table. It means we need to introduce a SyncResponse class mapped to a
sync_response table which this route will process."_

lccpinto replied:
I got it... I should have renamed the file **request-consumer** to **sender-message-consumer** instead, and then incorporate the new code from **sync-response-consumer**, in place of deleting **request-consumer** and use the **sync-response-consumer** file. This way, the git history would be more concise, with a renamed file rather than deleted.
My bad.

wluyima commented:
1. There is conflicts you need to resolve

wluyima commented on distribution/sender/routes/active-mq-publisher.xml:
```diff
@@ -0,0 +1,132 @@
+<routes xmlns="http://camel.apache.org/schema/spring">
+
+    <route id="active-mq-publisher" errorHandlerRef="outBoundErrorHandler">
```
I'm thinking the error handler for this route or the out-bound-db-sync should be different otherwise if an error occurs in this route, the retry mechanism might fail or might take it through the previous routes which you might not want

wluyima replied:
We need to think about this carefully otherwise we might end up in circles in case of errors, I will think through it, did you actually test the functionality including if an error occurs? And see what happens?

wluyima replied:
Please use the **shutdownErrorHandler** here otherwise the **outBoundErrorHandler** won't work when it is invoked in case of an exception in this route

wluyima replied:
Did you see these comments?

lccpinto replied:
I've changed to the **shutdownErrorHandler** as you suggested. Thanks.

wluyima reviewed (changes requested):
I added comments

wluyima commented on distribution/sender/routes/sender-message-consumer.xml:
```diff
@@ -0,0 +1,80 @@
+<routes xmlns="http://camel.apache.org/schema/spring">
+    <route id="sender-message-consumer" errorHandlerRef="shutdownErrorHandler">
```
Why did you renamed this?

wluyima commented on distribution/sender/routes/sender-message-consumer.xml:
```diff
@@ -0,0 +1,80 @@
+            
+            <otherwise>
+                <!-- TODO Log to a special failures log file or DB -->
+                <log message="Unknowk message was received: ${body}" loggingLevel="ERROR" />
```
You have a typo here for Unknown

wluyima replied:
This messages should be logged at warn not error

lccpinto replied:
Fixed.

wluyima commented on distribution/sender/routes/sync-response-processor.xml:
```diff
@@ -0,0 +1,66 @@
+            
+                            <toD uri="jpa:SenderSyncMessage?query=DELETE FROM SenderSyncMessage WHERE messageUuid = '${exchangeProperty.syncResponse.messageUuid}'" />
+            
+                            <log loggingLevel="DEBUG" message="Successfully removed Sender sync message(s) with uuid ${exchangeProperty.syncResponse.messageUuid}" />
```
I would rather have this message logged at info and the one on line 33 at debug

lccpinto replied:
Fixed

wluyima commented on distribution/sender/routes/sender-message-consumer.xml:
```diff
@@ -0,0 +1,80 @@
+<routes xmlns="http://camel.apache.org/schema/spring">
+    <route id="sender-message-consumer" errorHandlerRef="shutdownErrorHandler">
```
How about naming this sender-activemq-consumer or sender-activemq-message-consumer

lccpinto replied:
Renamed to sender-activemq-consumer.

wluyima commented on distribution/sender/routes/active-mq-publisher.xml:
```diff
@@ -0,0 +1,132 @@
+<routes xmlns="http://camel.apache.org/schema/spring">
+
+    <route id="active-mq-publisher" errorHandlerRef="outBoundErrorHandler">
```
Renamed the route id to sender-activemq-publisher

lccpinto replied:
Renamed.

wluyima commented:
I think this is a good starting point, can you please resolve the conflicts, we need to modify the retry mechanism because I think it is broken after your changes. I shared an a diagram of what I envision to be the new workflow that supports proper error handling.

wluyima commented:
@lccpinto please resolve the conflicts and merge the latest changes in master into you branch

lccpinto commented:
> @lccpinto please resolve the conflicts and merge the latest changes in master into you branch

Hi @wluyima, I've resolved the conflicts and merged master into this branch.

Commits:
0287db92b1fff0df780db307c085db56c3fd6d9d
5ea6cf613acb2f68df9bb1c365696cf65cd9f99e

wluyima commented:
@lccpinto there is still one unaddressed comment on the PR, and can you please resolve the conflicts?

lccpinto commented:
Hi @wluyima, I fixed the unaddressed comment.
But, I've resolved the conflicts, and now I see the message "This branch has no conflicts with the base branch
Merging can be performed automatically. ". Could you please tell what conflict is pending?
Thanks.

lccpinto commented:
Hi @wluyima, as we defined in last centralization meeting, I've implemented the EC-89 in this branch. Could you please review it?

Commit: 970b2d65495ce98e1c57c918dcba51769714dffe

Thanks

wluyima commented:
@lccpinto can you please resolve the conflicts?

wluyima commented on distribution/sender/routes/sync-response-processor.xml:
```diff
@@ -0,0 +1,66 @@
+
+        <from uri="scheduler:sync_response_processor?initialDelay={{sync-response-processor.initial.delay}}&amp;delay={{sync-response-processor.delay}}" />
+
+        <log message="Fetching next {{sync-response-processor.max.fetch.results}} sync responses" loggingLevel="DEBUG" />
```
Please use 1000 and remove this sync-response-processor.max.fetch.results property

lccpinto replied:
Changed to 500, as mentioned in [this comment](https://github.com/FriendsInGlobalHealth/openmrs-eip/pull/11#discussion_r869835394), and property was removed.

wluyima commented on app/src/main/java/org/openmrs/eip/app/utils/ReceiverMessageUtils.java:
```diff
@@ -0,0 +1,91 @@
+
+@Component
+@Profile(SyncProfiles.RECEIVER)
+public class ReceiverMessageUtils {
```
This should be a camel processor, i.e. a class that implements Processor, can you change it?

wluyima replied:
Camel has a built-in way to call processor with the processor tag

wluyima replied:
Now I see you are calling this from SiteMessageConsumer which is not a route which is why you made it a bean, I guess it is fine

wluyima replied:
But I find ReceiverMessageUtils strange for a bean class, to me it seems similar to the sender activemq publisher, how about ReceiverActiveMqMessagePublisher? Or ReceiverActiveMqMessageSender?

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:
```diff
@@ -0,0 +1,142 @@
+	
+	@Column(name = "error_message", length = 200)
+	@Access(AccessType.FIELD)
+	private String errorMessage;
```
Why do you have the error_message column? It is not supposed to be there

lccpinto replied:
Before this implementation there was a TODO in the sender/dbsync-route.xml. See the version of the commit 8667f83535be2c1dad29587224f4c4c5a147f988:

```
            <otherwise>
                <!-- TODO Log to a special failures log file or DB -->
                <log message="No entity found in the database matching identifier: ${exchangeProperty.event.identifier}" loggingLevel="ERROR" />
            </otherwise>
```

So, in order to implement this TODO, we've added this column that is filled in the sender-activemq-publisher route, when no entity is found:

```
                        <otherwise>
                            <setProperty name="errorMessage">
                                <constant>No entity found in the database matching identifier</constant>
                            </setProperty>
                            
                            <log message="${exchangeProperty.errorMessage}: ${exchangeProperty.senderSyncMessage.identifier}" loggingLevel="ERROR" />
                            <script>
                                <spel>
                                    #{getProperty('senderSyncMessage').markAsError(getProperty('errorMessage'))}
                                </spel>
                            </script>
                        </otherwise>
```

So, if this column is not supposed to be there, please tell what we should do for the SenderSyncMessages that during the process in the sender-activemq-publisher, does not have the correspondent entity in the database for the identifier. Then I can implement accordingly.

lccpinto replied:
Resolved in [this discussion](https://github.com/FriendsInGlobalHealth/openmrs-eip/pull/11#discussion_r905321859)

wluyima commented on app-api/src/main/resources/liquibase-sender.xml:
```diff
@@ -250,5 +250,74 @@
+                <constraints nullable="false" />
+            </column>
+            <column name="date_updated" type="DATETIME" />
+            <column name="error_message" type="VARCHAR(200)" />
```
Please remove this error_message column

lccpinto replied:
There was a purpose behind the creation of this column. I think that the better approach is to try to understand it before asking for removal. If this is not the correct way of doing things and after the explanation the decision to remove remains, then we will remove.

Please see [this comment](https://github.com/FriendsInGlobalHealth/openmrs-eip/pull/11#discussion_r907182671)

wluyima replied:
It is not an error if the entity is not found, the column is not required and currently the original logic is broken when I tested because the receiver is not getting the response back to resolve the sync request so it stays hanging forever, please don't change the original code, let's not introduced an new unnecessary functionalities or columns for as part of this work

wluyima replied:
I added a comment to respond to you, the TODO is for something else

lccpinto replied:
Like I said before, I did not changed the logic for sync requests. When no entity is found we still send an empty response, as you can see here:
https://github.com/FriendsInGlobalHealth/openmrs-eip/blob/faa9961347844922604c8f13d3ebd77550560298/distribution/sender/routes/sender-activemq-publisher.xml#L66-L80

But my question remains:
When we receive a database event, we create a SenderSyncMessage with the status 'NEW'.
Then, sender-activemq-publisher processes all the SenderSyncMessage with the status 'NEW'.
So, when we have a SenderSyncMessage from a regular database change (not a sync request) for which we do not found the correspondent entity, what do we do? Do we change the status or delete it? Otherwise it will be reprocessed in the sender-activemq-publisher forever.

wluyima replied:
If we don't find an entity for a database event, usually this is because the entity was deleted by another event and it is not an error, we just log the message, ignore that TODO, in fact remote it.

wluyima replied:
Delete the sync message row if the entity is not found when processing a regular DB event

wluyima commented on app-api/src/main/resources/liquibase-sender.xml:
```diff
@@ -250,5 +250,74 @@
+                <constraints nullable="false" unique="true"/>
+            </column>
+            <column name="request_uuid" type="VARCHAR(38)" />
+            <column name="date_created" type="DATETIME">
```
Datatype should be datetime(3)

lccpinto replied:
Changed.

wluyima commented on app-api/src/main/resources/liquibase-sender.xml:
```diff
@@ -250,5 +250,74 @@
+            <column name="status" type="VARCHAR(20)">
+                <constraints nullable="false" />
+            </column>
+            <column name="date_updated" type="DATETIME" />
```
Datatype should be datetime(3) and column name should be date_changed, this is the name we use conventionally

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:
```diff
@@ -0,0 +1,142 @@
+	
+	@Column(name = "date_updated")
+	@Access(AccessType.FIELD)
+	private Date dateUpdated;
```
Rename to dateChanged and column name to date_changed

lccpinto replied:
Done.

wluyima commented on distribution/sender/routes/sender-activemq-publisher.xml:
```diff
@@ -0,0 +1,149 @@
+                            <log message="${exchangeProperty.errorMessage}: ${exchangeProperty.senderSyncMessage.identifier}" loggingLevel="ERROR" />
+                            <script>
+                                <spel>
+                                    #{getProperty('senderSyncMessage').markAsError(getProperty('errorMessage'))}
```
**SenderSyncMessage** can't have error state, if something fails the applications should stop which why you set the error handler to shutdownErrorHandler

wluyima replied:
Please don't change the logic, if no entity is found, you do nothing just like the old code

wluyima replied:
Please fix this code to behave like before in master, you should send send the message back to the receiver with no entity payload, don't change the status to ERROR, infact remove the ERROR status from the status enum and remove the error_message field.

lccpinto replied:
I think that there is a misunderstand here. Lets go though each point:

_"SenderSyncMessage can't have error state, if something fails the applications should stop which why you set the error handler to shutdownErrorHandler"_
The only error that will be in this column for now is the one that was only logged, and that has a TODO to persist it in the DB (see [this comment](https://github.com/FriendsInGlobalHealth/openmrs-eip/pull/11#discussion_r907182671)). All other errors will remain lead to stop the application. 

_"Please don't change the logic, if no entity is found, you do nothing just like the old code"_
There is two scenarios where we can face this, which have different flows:
1. if we do not found the entity for Sync Requests from Central, we send an empty payload. This logic did not changed.

2. if we do not found the entity for all the other events, like create/update/delete, then we do not send an empty payload. In the previous version, we only logged this error and there was a TODO (see this 8667f83535be2c1dad29587224f4c4c5a147f988). And now, since we have a SenderSyncMessage register in the database, we cannot only log. For now, we mark it as ERROR to avoid reprocessing, and write the error in the column.
If this error information must not be there, we need to decide what to do with this SenderSyncMessages:
2.1. Mark with ERROR only, without error message
2.2. Simply log it like before, and remove the SenderSyncMessage to avoid reprocessing.

_"Please fix this code to behave like before in master, you should send send the message back to the receiver with no entity payload, don't change the status to ERROR, infact remove the ERROR status from the status enum and remove the error_message field."_
Please see the previous comment.

wluyima replied:
When processing a regular non delete DB event(insert/update) i.e. one where operation property is not set to **r** and the entity is not found, it means there is nothing to send the receiver, just log the message and delete the sync message row.

lccpinto replied:
Implemented accordingly.

wluyima reviewed (commented):
Added more comments

wluyima commented:
And please pull the latest changes from master into your branch

wluyima commented on distribution/sender/routes/sender-activemq-consumer.xml:
```diff
@@ -0,0 +1,93 @@
+        <setProperty name="requestUuid">
+            <jsonpath suppressExceptions="true">$.requestUuid</jsonpath>
+        </setProperty>
+        <setProperty name="messageUuid">
```
Isn't this a reputation?

wluyima commented on distribution/sender/routes/sender-activemq-consumer.xml:
```diff
@@ -0,0 +1,93 @@
+        <log message="Received message -> ${body}" />
+        
+        <setProperty name="requestUuid">
+            <jsonpath suppressExceptions="true">$.requestUuid</jsonpath>
```
Please don't suppress expressions otherwise you can mask an actual issue

wluyima replied:
What if the requestUuid exists but the unmarshalling fails because of some other issue? Don't you think it would you are masking a problem and might process the message wrongly?

wluyima commented on distribution/sender/routes/sender-activemq-consumer.xml:
```diff
@@ -0,0 +1,93 @@
+            </when>
+            
+            <when>
+                <simple>${exchangeProperty.messageUuid} != null</simple>
```
This is the same condition as the one in the first when clause, why?

wluyima commented on distribution/sender/routes/sender-activemq-consumer.xml:
```diff
@@ -0,0 +1,93 @@
+
+        <log message="Received message -> ${body}" />
+        
+        <setProperty name="requestUuid">
```
requestUuid only exists for a sync messages when responding to a sync request, so you should only expect requestUuid to exist if operation.operation is set to **r** in the incoming message, so rather change line 16 below to use check for this rather than doing here to avoid using suppressExceptions

wluyima commented on distribution/sender/routes/sender-activemq-consumer.xml:
```diff
@@ -0,0 +1,93 @@
+        
+        <choice>
+            <when>
+                <simple>${exchangeProperty.requestUuid} != null</simple>
```
I think here you should instead check if operation.operation is not set to **r** on the incoming message.

wluyima commented on distribution/receiver/routes/receiver-route.xml:
```diff
@@ -91,12 +91,18 @@
 
                     <log message="Successfully saved updates for sync request" loggingLevel="DEBUG" />
+                    
+                    <setProperty name="receiverSyncRequest">
```
Move the code so that you store the sync request in the property before saving it to the database

wluyima commented on camel-openmrs/src/main/java/org/openmrs/eip/component/utils/DateUtils.java:
```diff
@@ -1,53 +1,61 @@
-        return latestDateFromColl2 == null || latestDateFromColl1.isAfter(latestDateFromColl2);
-    }
-
+	
```
What did you change in this class?

lccpinto replied:
When processing SenderSyncMessage responses, for converting the payload dateSent. See [this](https://github.com/FriendsInGlobalHealth/openmrs-eip/blob/faa9961347844922604c8f13d3ebd77550560298/distribution/sender/routes/sender-activemq-consumer.xml#L59-L61)

wluyima replied:
You actually don't need this method, see the comment on in the route where you use this method

lccpinto replied:
Right. Removed while addressing [this comment](https://github.com/FriendsInGlobalHealth/openmrs-eip/pull/11#discussion_r908923725).

wluyima commented on distribution/docs/receiver/application.properties:
```diff
@@ -169,6 +169,11 @@ spring.artemis.password=
+
 
 # *********************** PGP configuration if encryption is necessary (OPTIONAL) **************************************
 #
```
Please add the loggers for the new routes too and if you renamed any new routes update their loggers to match the new route names

lccpinto replied:
Ok, but no routes was created nor renamed in the receiver.

wluyima commented on distribution/docs/sender/application.properties:
```diff
@@ -37,6 +37,24 @@ db-event.retry.initial.delay=120000
+# Maximum results of each fetch of sender sync responses to be read by sync response processor route.
+sync-response-processor.max.fetch.results=500
 # ----------------------------------------------------------------------------------------------------------------------
 
```
Please add the loggers for the new routes too and if you renamed any new routes update their loggers to match the new route names

wluyima reviewed (commented):
Added more comments

wluyima commented on app/src/main/java/org/openmrs/eip/app/utils/ReceiverMessageUtils.java:
```diff
@@ -0,0 +1,91 @@
+	private static final Logger log = LoggerFactory.getLogger(ReceiverMessageUtils.class);
+	
+	@Value("${" + Constants.PROP_CAMEL_OUTPUT_ENDPOINT + "}")
+	private String endpointConfig;
```
Don't make this a property, just hard code like we do for the sender

wluyima replied:
Yes I would like to give admins as much power to configure things but we also need to sort of limit it, the more configuration  options they have the more errors they are likely to make

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncResponse.java:
```diff
@@ -0,0 +1,80 @@
+	
+	public static final long serialVersionUID = 1;
+	
+	public enum SyncResponseStatus {
```
I don't think this class should have a status, after a response is processed, delete it instead

lccpinto replied:
Implemented.

wluyima commented on distribution/sender/routes/sync-response-processor.xml:
```diff
@@ -0,0 +1,66 @@
+
+                    <script>
+                        <spel>
+                            #{getProperty('syncResponse').markAsProcessed()}
```
After a response is processed, delete it instead, there is no reason for keeping it around

lccpinto replied:
Implemented.

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:
```diff
@@ -0,0 +1,142 @@
+	public enum SenderSyncMessageStatus {
+		NEW,
+		SENT,
+		ERROR
```
The ticket does not say ERROR is one of the statuses, please remove it

lccpinto replied:
Removed.

wluyima commented on distribution/sender/routes/sender-activemq-publisher.xml:
```diff
@@ -0,0 +1,149 @@
+                                <constant>No entity found in the database matching identifier</constant>
+                            </setProperty>
+                            
+                            <log message="${exchangeProperty.errorMessage}: ${exchangeProperty.senderSyncMessage.identifier}" loggingLevel="ERROR" />
```
It's not error when the entity is not found, so log this at debug level

wluyima reviewed (commented):
Added more comments

wluyima commented on distribution/sender/routes/sender-activemq-publisher.xml:
```diff
@@ -0,0 +1,149 @@
+            
+                    <choice>
+                        <when>
+                            <simple>${body} != null || ${exchangeProperty.event.operation} == 'r'</simple>
```
This condition looks incorrect, there is no longer an exchange property named event, so this logic is broken

wluyima reviewed (commented):
Added more comments

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/utils/ReceiverActiveMqMessagePublisher.java:
```diff
@@ -0,0 +1,91 @@
+		String endpoint = getCamelOutputEndpoint(siteIdentifier);
+		String payload = JsonUtils.marshall(syncResponse);
+		
+		log.info("Sending Sync Response. Payload: {}; Endpoint: {}", payload, endpoint);
```
I think this should be logged at debug level

lccpinto replied:
Done, and removed the endpoint as requested [here](https://github.com/FriendsInGlobalHealth/openmrs-eip/pull/11#discussion_r905468552)

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/utils/ReceiverActiveMqMessagePublisher.java:
```diff
@@ -0,0 +1,91 @@
+		producerTemplate.sendBody(endpoint, payload);
+		
+		if (log.isDebugEnabled()) {
+			log.debug("Sync Response sent.");
```
And think this should be logged at info level as **Sync response sent for message with uuid: ${messageUuid}**

lccpinto replied:
Done.

wluyima reviewed (commented):
Added more comments

wluyima reviewed (commented):
Added more comments

lccpinto commented on distribution/sender/routes/sender-activemq-publisher.xml:
```diff
@@ -0,0 +1,149 @@
+    
+        <from uri="scheduler:active_mq_publisher?initialDelay={{active-mq-publisher.initial.delay}}&amp;delay={{active-mq-publisher.delay}}" />
+
+        <log message="Fetching next {{active-mq-publisher.max.fetch.results}} sender sync messages" loggingLevel="DEBUG" />
```
Changed to 500, as mentioned in [this comment](https://github.com/FriendsInGlobalHealth/openmrs-eip/pull/11#discussion_r869835394).

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncResponse.java:
```diff
@@ -0,0 +1,49 @@
+	@NotNull
+	@Column(name = "message_uuid", length = 38, nullable = false, updatable = false)
+	@Access(AccessType.FIELD)
+	private String messageUuid;
```
Why is the access type set to field? This field has a public setter, I think you can remove the @Access annotation

lccpinto replied:
This was put there when I was using a constructor to set the messageUuid once, and at that time the setter did not exist.
But then you told me to remove the constructor and this annotation was left by mistake after creating the setter. I will remove it.

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncResponseModel.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Data;
+
+@Data
+public class SenderSyncResponseModel {
```
Name should be SyncResponseModel because it is not specific to the sender

lccpinto replied:
Alright. Renamed.

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncResponse.java:
```diff
@@ -0,0 +1,49 @@
+	@NotNull
+	@Column(name = "date_sent", nullable = false, updatable = false)
+	@Access(AccessType.FIELD)
+	private Date dateSent;
```
Same, here why is the access type set to field? This field has a public setter, I think you can remove the @access annotation

lccpinto replied:
The explanation is the same as [here](https://github.com/FriendsInGlobalHealth/openmrs-eip/pull/11#discussion_r909244325).

I'll remove it.

wluyima commented on app-api/src/main/resources/liquibase-sender.xml:
```diff
@@ -250,5 +250,69 @@
+            <column name="snapshot" type="BOOLEAN">
+                <constraints nullable="false" />
+            </column>
+            <column name="status" type="VARCHAR(20)">
```
I know am being picky here, apologies! But by unwritten convention we have been using VARCHAR(50) for these status columns, let's try to keep it uniform

lccpinto replied:
Alright, I'll take more attention to look for the existing column definitions when creating similar ones.
Changed to 50.

wluyima commented on distribution/sender/routes/sender-activemq-publisher.xml:
```diff
@@ -0,0 +1,146 @@
+                            <log message="Successfully updated sender sync message with identifier ${exchangeProperty.senderSyncMessage.identifier} to ${exchangeProperty.senderSyncMessage.status}" />
+                        </when>
+                        
+                        <otherwise>
```
I think you need a check here for if body is null and ${exchangeProperty. senderSyncMessage.operation} == 'r'

lccpinto replied:
This was clarified during our call.

wluyima commented on distribution/sender/routes/sender-activemq-publisher.xml:
```diff
@@ -0,0 +1,146 @@
+                        </when>
+                        
+                        <otherwise>
+                            <log message="No entity found in the database matching identifier ${exchangeProperty.senderSyncMessage.identifier}" loggingLevel="DEBUG" />
```
I think this should explicitly say **Requested entity not found in the database matching identifier ${exchangeProperty.senderSyncMessage.identifier} in table ${exchangeProperty.senderSyncMessage.tableName}**

wluyima replied:
And you need to log the message at info level like before, not debug

lccpinto replied:
Alright. Done.

lccpinto replied:
During our call it was decided to change this message to: 
`No entity found in the database matching identifier ${exchangeProperty.senderSyncMessage.identifier} in table ${exchangeProperty.senderSyncMessage.tableName}`

wluyima commented on distribution/sender/routes/sender-activemq-consumer.xml:
```diff
@@ -0,0 +1,93 @@
+                </setProperty>
+                
+                <setProperty name="dateSent">
+                    <method beanType="org.openmrs.eip.component.utils.DateUtils" method="localDateTimeToDate(${exchangeProperty.senderSyncResponseModel.dateSent})" />
```
Why do you need to convert this date, just copy the already converted dateSent value from the senderSyncResponseModel you just unmarshalled above

wluyima commented on distribution/sender/routes/sender-activemq-consumer.xml:
```diff
@@ -0,0 +1,93 @@
+                </setProperty>
+                
+                <setProperty name="dateSent">
+                    <method beanType="org.openmrs.eip.component.utils.DateUtils" method="localDateTimeToDate(${exchangeProperty.senderSyncResponseModel.dateSent})" />
```
You don't need this, just change the field type in SenderSyncResponse class to LocalDateTime instead of Date that way you don't need this conversion

lccpinto replied:
Alright, done.

wluyima commented on app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncResponse.java:
```diff
@@ -0,0 +1,49 @@
+	
+	@NotNull
+	@Column(name = "date_sent", nullable = false, updatable = false)
+	@Access(AccessType.FIELD)
```
The value of this date field is parsed from the payload sent by the receiver, I suggest you change this type from Date to LocalDateTime so that you don't have to add and use the new method in DateUtils

lccpinto replied:
Changed accordingly.

lccpinto replied:
The field has been changed to dateSentByReceiver, as we discussed.

wluyima commented on distribution/sender/routes/sync-response-processor.xml:
```diff
@@ -0,0 +1,58 @@
+                <simple>${body.size()} > 0</simple>
+                <log message="Fetched ${body.size()} sender sync responses" />
+                
+                <split parallelProcessing="false">
```
Set stopOnException to true

wluyima commented on distribution/sender/routes/sender-activemq-publisher.xml:
```diff
@@ -0,0 +1,146 @@
+                <simple>${body.size()} > 0</simple>
+                <log message="Fetched ${body.size()} sender sync messages" />
+
+                <split parallelProcessing="false">
```
Set stopOnException to true

wluyima commented on camel-openmrs/src/main/java/org/openmrs/eip/component/utils/DateUtils.java:
```diff
@@ -7,47 +7,49 @@
+	 * @return returns true ONLY if dates1 contains the latest non null date instance otherwise
+	 *         false
+	 */
+	public static boolean containsLatestDate(Collection<LocalDateTime> dates1, Collection<LocalDateTime> dates2) {
```
You forgot to remove this method, because I believe you no longer use it anywhere, just revert everything in this file I guess.

lccpinto replied:
Done, file reverted.

wluyima commented on distribution/docs/sender/application.properties:
```diff
@@ -37,6 +37,18 @@ db-event.retry.initial.delay=120000
 events.sync.parallel.size=
+
+# The interval in milliseconds before the first sender sync messages fetch by the active mq publisher route when the application starts.
+active-mq-publisher.initial.delay=120000
```
2 minutes is a long time for initial delay

lccpinto replied:
Wyclif. 
In the first version, I set this delay to 15 seconds. Later on, in [this comment](https://github.com/FriendsInGlobalHealth/openmrs-eip/pull/11#discussion_r869834783), you suggested to change it to two minutes, to give time to other routes with more priority...

Now I'm confused seeing you saying that the two minutes that you said to be used is a long time!

wluyima commented on distribution/docs/sender/application.properties:
```diff
@@ -37,6 +37,18 @@ db-event.retry.initial.delay=120000
+active-mq-publisher.delay=60000
+
+# The interval in milliseconds before the first sender sync responses fetch by the sync response processor route when the application starts.
+sync-response-processor.initial.delay=120000
```
2 minutes is a long time for initial delay

lccpinto replied:
What delay should we set?
